src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs

```diff
@@ -423,6 +423,75 @@ checkLiteral = \case
   LitQName  _   -> return ()
   LitMeta   _ _ -> return ()
 
+-- | Check a `record where` expression.
+checkRecordWhere :: Range
+                 -> [C.Declaration]
+                 -> (RecInfo -> A.Assigns -> A.Expr)
+                 -> ScopeM A.Expr
+checkRecordWhere r ds0 mkRec = do
+  let i  = ExprRange r
+      ri = recInfoWhere r
+  -- We need to rewrite any opens or module applications opened in this
+  -- scope into a non-opening declaration and explicit bindings, so that
+  -- references inside this body refer unambiguously to things opened
+  -- here. Conveniently, this also homogenizes the things that need to be
+  -- made fields in the final `A.Rec`. If we get #3801, perhaps this
+  -- rewrite won't be necessary any more, but any `using` or `renaming`
+  -- clauses will still have to be made into field assignments somehow.
+  ds0' <- mapM' rewriteConcreteOpens ds0
+  List1.ifNull ds0' (return $ A.Rec ri []) $ \ ds -> do
+    localToAbstract (LetDefs ds) $ \ ds' -> do
+      names <- forM' ds' $ \case
+        A.LetBind _ _ (A.BindName name) _ _ -> return [name]
+        A.LetPatBind _ p _ -> go p
+          where
+            go = \case
+              A.VarP (A.BindName name)    -> return [name]
+              A.ConP _ _ args             -> mapM' (go . namedArg) args
+              A.RecP _ fs                 -> mapM' (go . _exprFieldA) fs
+              A.WildP _                   -> return []
+              A.AsP _ (A.BindName name) p -> (name :) <$> go p
+              A.PatternSynP _ _ args      -> mapM' (go . namedArg) args
+              -- We have run checkValidLetPattern at this point
+              A.ProjP{}   -> __IMPOSSIBLE__
+              A.DefP{}    -> __IMPOSSIBLE__
+              A.DotP{}    -> __IMPOSSIBLE__
+              A.AbsurdP{} -> __IMPOSSIBLE__
+              A.LitP{}    -> __IMPOSSIBLE__
+              A.EqualP{}  -> __IMPOSSIBLE__
+              A.WithP{}   -> __IMPOSSIBLE__
+        _ -> return []
+      let fs = [C.FieldAssignment (nameCanonical n) (A.Var n) | n <- names]
+      return $ A.mkLet i ds' (mkRec ri fs)
+  where
+    rewriteConcreteOpens :: C.Declaration -> ScopeM [C.Declaration]
+    rewriteConcreteOpens = \case
+      C.Open _ m dir -> handleOpen m dir
+      C.ModuleMacro r e m mapp DoOpen dir -> do
+        m' <- case m of
+          NoName r _ -> freshConcreteName r 0 ".modulemacro"
+          n -> return n
+        ds <- handleOpen (C.QName m') dir
+        return $ C.ModuleMacro r e m' mapp DontOpen defaultImportDir : ds
+
+      C.Private r o ds -> singleton . C.Private r o <$> mapM' rewriteConcreteOpens ds
+      d -> return [d]
+
+    handleOpen :: C.QName -> C.ImportDirective -> ScopeM [C.Declaration]
+    handleOpen m ImportDirective{ using, impRenaming } = case using of
+      UseEverything ->
+        typeError OpenEverythingInRecordWhere
```

Comment:
What's the rationale here? We do allow `record { Foo }` (where `Foo` is a module), so I'd expect this to also be supported.

Reply:
This was in the original PR and I think the motivation was this technicality:

https://github.com/agda/agda/blob/3b8686686f4a665b5be0019d382fd4dac21b0a72/src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs#L434-L440

(#3801 is the suggestion to allow more shadowing)

It's kind of important to allow outer bindings to be shadowed in a `record where`, since a common use is to open another instance of the record, and this would create ambiguous overloaded projections without the shadowing.

From a user perspective I guess you could argue that there is a difference in that in `record { Foo }` you're explicitly saying "here's a module that defines the fields of the record", whereas in a `record where` you might open modules to get access to other functions to build the fields.

I think it would make sense to lift the restriction, but it would be much easier to do after #3801 is fixed.